datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -345,34 +345,37 @@ impl CaseExpr {
         let when_expr = &self.when_then_expr[0].0;
         let then_expr = &self.when_then_expr[0].1;
 
-        let when_expr_value = when_expr.evaluate(batch)?;
-        let when_expr_value = match when_expr_value {
+        match when_expr.evaluate(batch)? {
+            // WHEN true --> column
+            ColumnarValue::Scalar(ScalarValue::Boolean(Some(true))) => {
+                then_expr.evaluate(batch)
+            }
+            // WHEN [false | null] --> NULL
             ColumnarValue::Scalar(_) => {
-                ColumnarValue::Array(when_expr_value.into_array(batch.num_rows())?)
```

Comment:
This line has the effect of creating an array out of a scalar -- I remove that and added the special case handling for the values